feat(flags): Flag evaluation support for flags that depend on other flags

haacked opened the pull request:
Remember all those refactoring PRs. Well this is what they were leading to!

[!IMPORTANT]
> 👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Problem

This is the backend support for #29016. It does not include the UI for creating such flags. This enables us to have flags that depend on other flag evaluation results.

The next step is for me to update the UI to support creating such flags.

## Changes

Builds a dependency graph of flags and then evaluates flags by level.

## Did you write or update any docs for this change?

<!-- Engineers are responsible for doing the first pass at documenting their features and/or code.  -->

- [ ] I've [added or updated the docs](https://posthog.com/handbook/engineering/writing-docs)
- [ ] I've reached out for help from the docs team
- [x] No docs needed for this change

## How did you test this code?

* Unit tests
* Manual tests via Postman

haacked committed:
Add unit tests for feature flag dependency evaluation

Also add test for cycles and missing dependencies and ensure the database is only called once.

dmarticus commented on rust/feature-flags/src/flags/flag_matching.rs:
```diff
@@ -420,16 +421,42 @@ impl FeatureFlagMatcher {
+        let (flag_dependency_graph, errors) =
+            match DependencyGraph::from_nodes(&feature_flags.flags) {
+                Ok((graph, errors)) => (graph, errors),
+                Err(e) => {
+                    error!("Failed to build feature flag dependency graph: {:?}", e);
+                    return FlagsResponse {
+                        errors_while_computing_flags: true,
+                        flags: HashMap::new(),
+                        quota_limited: None,
+                        request_id,
+                        config: ConfigResponse::default(),
+                    };
+                }
+            };
+
+        if !errors.is_empty() {
+            errors_while_computing_flags = true;
+            // TODO: Make sure this call reports a useful error message.
+            error!("There were errors building the feature flag dependency graph. Will attempt to evaluate the rest of the flags: {:?}", errors);
+        }
```
in both of these error paths, I would increment a metric counter (like [this](https://github.com/PostHog/posthog/blob/master/rust/feature-flags/src/flags/flag_service.rs#L58-L62) so we can scrape them on our [dashboard](https://grafana.prod-us.posthog.dev/d/eehu20mrbkbuof/new-feature-flags?var-reason=$__all&var-team_id=$__all&var-timing_quartile=0.99).  Each time I add metrics to this service I'm grateful I have them, it's super easy, and costs basically nothing at this point.

dmarticus commented on rust/feature-flags/src/properties/property_models.rs:
```diff
@@ -44,6 +44,9 @@ pub struct PropertyFilter {
     pub operator: Option<OperatorType>,
     #[serde(rename = "type")]
     pub prop_type: PropertyType,
+    // NB: a client-only property used to transform operators to their negation counterparts. For example, IsNot becomes Exact and vice versa.
+    // This is NOT used during flag property matching.
+    #[serde(skip_serializing_if = "Option::is_none")]
     pub negation: Option<bool>,
```
👍

dmarticus commented on rust/feature-flags/src/flags/flag_matching.rs:
```diff
@@ -420,16 +421,42 @@ impl FeatureFlagMatcher {
+                    hash_key_overrides.clone(),
+                )
+                .await;
+            errors_while_computing_flags |= level_errors;
```
hell yeah, the BitOrAssign operator

dmarticus commented on rust/feature-flags/src/flags/flag_matching.rs:
```diff
@@ -420,16 +421,42 @@ impl FeatureFlagMatcher {
+            match DependencyGraph::from_nodes(&feature_flags.flags) {
+                Ok((graph, errors)) => (graph, errors),
+                Err(e) => {
+                    error!("Failed to build feature flag dependency graph: {:?}", e);
```
```suggestion
                    error!("Failed to build feature flag dependency graph for team {}: {:?}", team_id, e);
```

dmarticus commented on rust/feature-flags/src/flags/flag_matching.rs:
```diff
@@ -420,16 +421,42 @@ impl FeatureFlagMatcher {
+        }
+
+        let mut flag_details_map = HashMap::new();
+        let evaluation_stages = flag_dependency_graph.evaluation_stages().unwrap();
+        for stage in evaluation_stages {
+            let stage_flags: Vec<FeatureFlag> = stage.iter().map(|&flag| flag.clone()).collect();
+            let (level_flag_details_map, level_errors) = self
+                .evaluate_flags_in_level(
+                    &stage_flags,
+                    &person_property_overrides,
+                    &group_property_overrides,
+                    hash_key_overrides.clone(),
+                )
+                .await;
```
this worries me because `evaluate_flags_in_level` is async and makes database calls via `prepare_flag_evaluation_state`.  If we have a lot of dependency levels, this is a lot of slow operations.  

Why aren't we making the database call before all of this?  At worst case we should only ever make one call per flag request to evaluate all of the properties of all of the flags that we need.  As it stands now this feels like an unacceptable performance hit, although perhaps my read of this is naive.

haacked replied:
> this worries me because `evaluate_flags_in_level` is async and makes database calls via prepare_flag_evaluation_state. If we have a lot of dependency levels, this is a lot of slow operations.

This is a valid concern. It will only make a db call per level if any flag in that level needs db properties to evaluate. But for a deeply nested dependency tree this could result in more db calls. I think in 99.9% of cases, the behavior will be exactly what we have before. But yeah, this is a concern.

> Why aren't we making the database call before all of this?

`evaluate_flags_in_level` is effectively what we do today (before this PR). The way it knows which flags need DB properties is it attempts to evaluate the flag first and if the result is `None`, it adds that flag to the `flags_needing_db_properties` vector.

I wasn't sure if there's a good way to determine ahead of time which flags need DB properties without evaluating them. I can't really evaluate all of them ahead of time because of the dependencies.

I'll dig a little deeper to see if I have any ideas.

haacked replied:
Ok, I resolved this in #34248.

dmarticus reviewed (commented):
I think this is functionally correct but I have some concerns about performance.

haacked committed:
Evaluate flags in stages

For flags with no dependencies, it will evaluate them all in a single stage.

But if there are flags with dependencies, it will evaluate them in stages from those with no dependencies to those with dependencies.

haacked committed:
Add a debug assertion

haacked committed:
Refactor flag evaluation with clearer error handling

Splits the flag evaluation process into distinct steps: group type mapping initialization, DB property preparation, and dependency graph evaluation. Introduces dedicated methods for each step, improves error handling and logging, and ensures errors are tracked and reported at each stage.

andyzzhao commented on rust/feature-flags/src/flags/flag_matching.rs:
```diff
@@ -425,62 +426,192 @@ impl FeatureFlagMatcher {
+            Ok(stages) => stages,
+            Err(e) => {
+                self.handle_dependency_graph_error("get evaluation stages", &e);
+                debug_assert!(
```
Why keep this `debug_assert`?

haacked replied:
This code path should be unreachable. And my normal preference is to indicate that by calling `unreachable` here. However, should I be wrong about that (for example, maybe someone changes something later that invalidates that contract), I don't want to bring down this service with a panic. I'd rather it fail gracefully.

A `debug_assert` feels like a nice compromise. When developing, we'll get a panic and it will be impossible to ignore. But in production, the service will go on. I'm happy to reconsider. I'll add a comment.

andyzzhao commented on rust/feature-flags/src/flags/flag_matching.rs:
```diff
@@ -425,62 +426,192 @@ impl FeatureFlagMatcher {
+            FLAG_EVALUATION_ERROR_COUNTER,
+            &[(
+                "reason".to_string(),
+                format!("{}_error", error_type.replace(" ", "_")),
```
nit: Unless someone is already aware of this formatting convention, it'll be hard to figure out where `build_feature_flag_dependency_graph_error` is coming from. They will search the code base for `build_feature_flag_dependency_graph_error`, and it will return no results, which is a bit painful, especially during an incident.

andyzzhao replied:
oh wait this is a metric, ignore me

andyzzhao reviewed (approved):
This was really easy to follow, thank you 🚀

haacked committed:
Add clarifying comment
